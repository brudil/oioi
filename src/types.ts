export enum NodeType {
  FunctionDeclaration = 'FunctionDeclaration',
  Identifier = 'Identifier',
  Integer = 'Integer',
  Operator = 'Operator',
  Expression = 'Expression',
  Statement = 'Statement',
}

export enum OperatorType {
  Plus = '+',
  Minus = '-',
  Multiply = '*',
  Div = '/'
}

export interface SourcePosition {
  offset: number;
  lineBreaks: number;
  line: number;
  col: number;
}

export interface ProducerData extends SourcePosition {
  type: string;
  value: string;
  text: string;
  [pos: number]: any;
}

export type NodeProducer<n> = (data: ProducerData) => n;

export interface ASTNode {
  type: NodeType;
  position: SourcePosition;
}

export interface MainNode extends ASTNode {
  program: ASTNode[];
}

export interface ExpressionNode extends ASTNode {
  lhs: any;
  rhs: any;
  operator: OperatorType;
}

export type Node = ExpressionNode;
