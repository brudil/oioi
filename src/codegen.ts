import * as binaryen from 'binaryen';
import { MainNode, ASTNode, NodeType, ExpressionNode } from './types';

export function generate(ast: MainNode) {
  const b = new binaryen.Module();

  function createOrGetFunctionType(rtn: binaryen.Type, params: binaryen.Type[]) {
    return b.addFunctionType("fntype", rtn, params);
  }

  function walkExp(_node: ExpressionNode) {

  }

  function walkAstNode(node: ASTNode) {
    switch (node.type) {
      case NodeType.Expression: {
        walkExp(node as ExpressionNode);
      }
    }
  }

  function walkMainNode(ast: MainNode) {
    ast.program.map(walkAstNode)
  }

  walkMainNode(ast);

  b.addFunction("rah", createOrGetFunctionType(binaryen.none, []), [binaryen.i32], b.block("rah2", [
    b.setLocal(2,
      b.i32.add(
        b.getLocal(0, binaryen.i32),
        b.getLocal(1, binaryen.i32)
      )
    ),
    b.return(
      b.getLocal(2, binaryen.i32)
    )
  ]));

  return b;
}
