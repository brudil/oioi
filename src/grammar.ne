@{%
  const lexer = require('../lexer.ts');

  const empty = () => null;

  const p = require('../producers.ts');
%}

@preprocessor typescript
@lexer lexer

Main -> Statement:+ {% d => ({ program: d }) %}

Statement -> Expression | Fn | Newlines {% d => p.createStatement(d[0]) %}

Expression -> Integer _ %OP  _ Integer {% d => ({"lhs": d[0], "rhs": d[4], "operator": d[2].value }) %}

Fn -> "fn" __ Identifier _ Block {% d => p.createFunction({ identifier: d[2], body: d[4] }) %}

Params -> null | Param | Params _ " " Param {% d => ({ "type": "params", d }) %}

Block -> %LBRACE _ Body _ %RBRACE {% d => d[2] %}

Body -> Statement:+ | null {% d => d %}

Param -> Integer

Identifier -> %Ident {% d => p.createIdentifier({ name: d[0].value}) %}

Integer -> %number {% p.createInteger %}

_N -> Newlines | _

Newlines -> %NL:+ {% empty %}

# Whitespace
_ -> null | %WS {% empty %}
__ -> %WS | __ %WS {% empty %}
