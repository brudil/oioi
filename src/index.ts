const nearley = require('nearley');
const grammar = require("./generated/grammar.js");

// Create a Parser object from our grammar.
const parser = new nearley.Parser(nearley.Grammar.fromCompiled(grammar));

// Parse something!
parser.feed("foo\n");

// parser.results is an array of possible parsings.
parser.results.map(console.log.bind(console));