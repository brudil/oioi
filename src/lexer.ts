const moo = require("moo");

const lexer = moo.compile({
  WS:     /[ \t]+/,
  number:  /0|[1-9][0-9]*/,
  string:  /"(?:\\["\\]|[^\n"\\])*"/,
  NL:      { match: /\n/, lineBreaks: true },
  times:  /\*|x/,
  OP: ['*', '+', '/', '-'],
  Ident: /[a-zA-Z_]+/,
  LBRACE: '{',
  RBRACE: '}',
});

module.exports = lexer;
