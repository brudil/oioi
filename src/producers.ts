import { NodeProducer, NodeType } from "./types";

export const createInteger: NodeProducer<number> = (d) => parseInt(d[0], 10);

export const createFunction = ({ identifier, body }: { identifier: any, body: any }) => ({ type: NodeType.FunctionDeclaration, identifier, body });

export const createIdentifier = ({ name }: { name: string }) => ({ type: NodeType.Identifier, name });

export const createStatement = (statement: any) => ({ type: NodeType.Statement, body: statement });
