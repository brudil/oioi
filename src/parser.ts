import { readFileSync } from 'fs';
import * as nearley from 'nearley';
import * as grammar from './generated/grammar';


class AmbiguousGrammarError extends Error {}

export function parserFromPath(path: string) {
    const nParser = new nearley.Parser(nearley.Grammar.fromCompiled(grammar), { keepHistory: true });

    try {
        const source = readFileSync(path, 'utf-8');
        nParser.feed(source);
        if (nParser.results.length !== 1) {
            throw new AmbiguousGrammarError();
        }
        return nParser.results[0];
    } catch(e) {
        console.log('parsing failed');
        console.log(e);
        throw new Error('source failed to parse');
    }
}
