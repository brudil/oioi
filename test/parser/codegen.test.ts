import { resolve, normalize } from 'path';
import glob from 'fast-glob';
import { parserFromPath } from '../../src/parser';
import { generate } from '../../src/codegen';
import {filename} from "../utils";

const entries = glob.sync<string>('test/parser/*.oi');

entries.forEach(entry => {
    test(filename(entry), () => {
        const absPath = normalize(resolve(process.cwd(), entry));
        const ast = parserFromPath(absPath);
        expect(generate(ast).emitText()).toMatchSnapshot();
    });
});
