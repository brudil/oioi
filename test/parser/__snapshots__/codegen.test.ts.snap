// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`simple-arithmetic 1`] = `
"(module
 (type $fntype (func))
 (func $rah (; 0 ;) (type $fntype)
  (local $0 i32)
  (block $rah2
   (set_local $2
    (i32.add
     (get_local $0)
     (get_local $1)
    )
   )
   (return
    (get_local $2)
   )
  )
 )
)
"
`;

exports[`simple-function 1`] = `
"(module
 (type $fntype (func))
 (func $rah (; 0 ;) (type $fntype)
  (local $0 i32)
  (block $rah2
   (set_local $2
    (i32.add
     (get_local $0)
     (get_local $1)
    )
   )
   (return
    (get_local $2)
   )
  )
 )
)
"
`;
